src/qibo/transpiler/router.py

```diff
@@ -40,6 +40,108 @@ def assert_connectivity(connectivity: nx.Graph, circuit: Circuit):
+                as the original but respects the hardware connectivity,
+                and list that maps logical to hardware qubits.
+        """
+
+        middle_qubit = self.middle_qubit
+        nqubits = max(circuit.nqubits, middle_qubit + 1)
+        # new circuit object that will be compatible with hardware connectivity
+        new = Circuit(nqubits)
+        # list to maps logical to hardware qubits
+        hardware_qubits = list(initial_layout.values())
+
+        for i, gate in enumerate(circuit.queue):
+            # map gate qubits to hardware
+            qubits = tuple(hardware_qubits.index(q) for q in gate.qubits)
+            if isinstance(gate, gates.M):
+                new_gate = gates.M(*qubits, **gate.init_kwargs)
+                new_gate.result = gate.result
+                new.add(new_gate)
+                continue
+
+            if len(qubits) > 2:
+                raise ConnectivityError(
+                    "Gates targeting more than two qubits are not supported."
+                )
+
+            elif len(qubits) == 2 and middle_qubit not in qubits:
+                # find which qubit should be moved
+                new_middle = _find_connected_qubit(
+                    qubits, circuit.queue[i + 1 :], hardware_qubits
+                )
+                # update hardware qubits according to the swap
+                hardware_qubits[middle_qubit], hardware_qubits[new_middle] = (
+                    hardware_qubits[new_middle],
+                    hardware_qubits[middle_qubit],
+                )
+                new.add(gates.SWAP(middle_qubit, new_middle))
+                # update gate qubits according to the new swap
+                qubits = tuple(hardware_qubits.index(q) for q in gate.qubits)
+
+            # add gate to the hardware circuit
+            if isinstance(gate, gates.Unitary):
+                # gates.Unitary requires matrix as first argument
+                matrix = gate.init_args[0]
+                new.add(gate.__class__(matrix, *qubits, **gate.init_kwargs))
+            else:
+                new.add(gate.__class__(*qubits, **gate.init_kwargs))
+        hardware_qubits_keys = ["q" + str(i) for i in range(5)]
+        return new, dict(zip(hardware_qubits_keys, hardware_qubits))
+
+
+def _find_connected_qubit(qubits, queue, hardware_qubits):
+    """Helper method for :meth:`qibo.transpiler.StarConnectivity`.
+
+    Finds which qubit should be mapped to hardware middle qubit
+    by looking at the two-qubit gates that follow.
+    """
+    possible_qubits = set(qubits)
+    for next_gate in queue:
+        if len(next_gate.qubits) == 2:
+            possible_qubits &= {hardware_qubits.index(q) for q in next_gate.qubits}
+            if not possible_qubits:
+                # freedom of choice
+                return qubits[0]
+            elif len(possible_qubits) == 1:
+                return possible_qubits.pop()
+    # freedom of choice
+    return qubits[0]
```

Comment:
This same function here is outside of the class but in the placer is a method of the class. Any particular reason why?

Reply:
Thank you for noticing this. I have left just the function outside the class in the router and used it for the StarConnectivityPlacer